### PR TITLE
Code clean-ups

### PR DESCRIPTION
**Pull request by vince62s**

### CODE REVIEW

**Comment by vince62s** [comment]
Main difference in perf comes from the use of the context manager:
`with sdpa_kernel([SDPBackend.EFFICIENT_ATTENTION]):` before the `scaled_dot_product_attention()`
The counterpart is that it takes longer to startup because of some recompilings.
Not sure what is the best.

For the rest, some code clean-ups.

Before this PR, EuroLLM-9B finetuning the estimator:
```
[2024-12-27 09:50:51,765 INFO] Step 10/ 4000; acc: 54.8; ppl: 24.84; xent: 3.21; aux: 0.599; lr: 3.33e-06; sents:    2560; bsz:  208/ 208/ 2; 1694/1694 tok/s;    157 sec;
[2024-12-27 09:51:39,617 INFO] Step 20/ 4000; acc: 55.0; ppl: 24.50; xent: 3.20; aux: 0.548; lr: 6.67e-06; sents:    2560; bsz:  210/ 210/ 2; 5625/5625 tok/s;    205 sec;
[2024-12-27 09:52:28,284 INFO] Step 30/ 4000; acc: 55.3; ppl: 23.62; xent: 3.16; aux: 0.451; lr: 1.00e-05; sents:    2560; bsz:  213/ 213/ 2; 5602/5602 tok/s;    253 sec;
[2024-12-27 09:53:17,060 INFO] Step 40/ 4000; acc: 55.2; ppl: 24.26; xent: 3.19; aux: 0.248; lr: 1.33e-05; sents:    2560; bsz:  211/ 211/ 2; 5545/5545 tok/s;    302 sec;
[2024-12-27 09:54:06,210 INFO] Step 50/ 4000; acc: 55.4; ppl: 23.51; xent: 3.16; aux: 0.110; lr: 1.67e-05; sents:    2560; bsz:  215/ 215/ 2; 5592/5592 tok/s;    351 sec;
[2024-12-27 09:54:54,319 INFO] Step 60/ 4000; acc: 54.8; ppl: 25.14; xent: 3.22; aux: 0.061; lr: 2.00e-05; sents:    2560; bsz:  206/ 206/ 2; 5481/5481 tok/s;    399 sec;
```
This PR:
```
[2024-12-27 10:20:55,554 INFO] Step 10/ 4000; acc: 54.8; ppl: 24.81; xent: 3.21; aux: 0.599; lr: 3.33e-06; sents:    2560; bsz:  208/ 208/ 2; 830/830 tok/s;    320 sec;
[2024-12-27 10:21:41,045 INFO] Step 20/ 4000; acc: 55.0; ppl: 24.47; xent: 3.20; aux: 0.548; lr: 6.67e-06; sents:    2560; bsz:  210/ 210/ 2; 5917/5917 tok/s;    366 sec;
[2024-12-27 10:22:27,125 INFO] Step 30/ 4000; acc: 55.3; ppl: 23.59; xent: 3.16; aux: 0.451; lr: 1.00e-05; sents:    2560; bsz:  213/ 213/ 2; 5916/5916 tok/s;    412 sec;
[2024-12-27 10:23:13,582 INFO] Step 40/ 4000; acc: 55.2; ppl: 24.22; xent: 3.19; aux: 0.248; lr: 1.33e-05; sents:    2560; bsz:  211/ 211/ 2; 5821/5821 tok/s;    458 sec;
[2024-12-27 10:24:00,311 INFO] Step 50/ 4000; acc: 55.3; ppl: 23.48; xent: 3.16; aux: 0.110; lr: 1.67e-05; sents:    2560; bsz:  215/ 215/ 2; 5881/5881 tok/s;    505 sec;
[2024-12-27 10:24:45,582 INFO] Step 60/ 4000; acc: 54.8; ppl: 25.10; xent: 3.22; aux: 0.061; lr: 2.00e-05; sents:    2560; bsz:  206/ 206/ 2; 5825/5825 tok/s;    550 sec;

```

Before this PR, Encoder-Decoder training:
```
[2024-12-27 09:59:18,407 INFO] Step 100/200000; acc: 13.6; ppl: 10780.07; xent: 9.29; aux: 0.000; lr: 6.72e-06; sents:  118220; bsz: 8506/10554/197; 26799/33251 tok/s;    190 sec;
[2024-12-27 10:01:04,435 INFO] Step 200/200000; acc: 17.8; ppl: 3006.98; xent: 8.01; aux: 0.000; lr: 1.34e-05; sents:  107813; bsz: 8515/10553/180; 48188/59718 tok/s;    296 sec;
[2024-12-27 10:02:51,181 INFO] Step 300/200000; acc: 19.5; ppl: 1181.01; xent: 7.07; aux: 0.000; lr: 2.02e-05; sents:  100017; bsz: 8583/10569/167; 48244/59408 tok/s;    403 sec;
```
This PR:
```
[2024-12-27 10:11:14,616 INFO] Step 100/200000; acc: 13.6; ppl: 10779.73; xent: 9.29; aux: 0.000; lr: 6.72e-06; sents:  118220; bsz: 8506/10554/197; 12455/15453 tok/s;    410 sec;
[2024-12-27 10:12:59,324 INFO] Step 200/200000; acc: 17.8; ppl: 3007.03; xent: 8.01; aux: 0.000; lr: 1.34e-05; sents:  107813; bsz: 8515/10553/180; 48796/60471 tok/s;    514 sec;
[2024-12-27 10:14:44,983 INFO] Step 300/200000; acc: 19.5; ppl: 1184.54; xent: 7.08; aux: 0.000; lr: 2.02e-05; sents:  100017; bsz: 8583/10569/167; 48740/60020 tok/s;    620 sec;

```

File: eole/predict/generator.py [hunk 1]
```diff
@@ -180,11 +178,14 @@ def _predict_batch_with_strategy(self, batch, decode_strategy, left_pad=True):
             dec_in = torch.cat((src, dec_in), 1)
             tgt_pad_mask = dec_in.eq(self._tgt_pad_idx).unsqueeze(1)  # [B, T_tgt]
             emb = self.model.tgt_emb(dec_in)
+            self.model.decoder._clear_cache()
```
Comment (francoishernandez) [inline eole/predict/generator.py:181]: I'm not 100% convinced this is clearer than having it in the decoder forward, but I may be wrong.

File: eole/predict/generator.py [hunk 2]
```diff
@@ -180,11 +178,14 @@ def _predict_batch_with_strategy(self, batch, decode_strategy, left_pad=True):
...
+            position_embeddings = self.model.rope.update(dec_in.size(1), step=0)
```
Comment (francoishernandez) [inline eole/predict/generator.py:182]: Same as in inference.py.

Also, not strictly linked to this PR but these duplicated changes lead to the question of the duplicated code -> shouldn't this estimator part be moved up in inference._decode_and_generate?

File: eole/predict/inference.py [hunk 3]
```diff
@@ -650,17 +653,17 @@ def _decode_and_generate(
...
         step=None,
-        batch_offset=None,
         return_attn=False,
     ):
 
         # Decoder forward, takes [batch, tgt_len, nfeats] as input
         # and [batch, src_len, hidden] as enc_out
         # in case of inference tgt_len = 1, batch = beam times batch_size
         # in case of Gold Scoring tgt_len = actual length, batch = 1 batch
+        emb = self.model.tgt_emb(decoder_in, step=step)
```
Comment (francoishernandez) [inline eole/predict/inference.py:665]: Specific reason for moving this?

File: eole/predict/inference.py [hunk 4]
```diff
@@ -674,8 +677,7 @@ def _decode_and_generate(
         else:
             src_pad_mask = None
         tgt_pad_mask = decoder_in.eq(self._tgt_pad_idx).unsqueeze(1)  # [B, 1, T_tgt]
-
-        emb = self.model.tgt_emb(decoder_in, step=step)
+        position_embeddings = self.model.rope.update(decoder_in.size(1), step=step)
```
Comment (francoishernandez) [inline eole/predict/inference.py:680]: I find this a bit weird to have it in the generic generator. In theory we could have some models that do not have position embeddings. (The no-op is there to cover this case, but it's a bit unclear.)

File: eole/predict/inference.py [hunk 5]
```diff
@@ -674,8 +677,7 @@ def _decode_and_generate(
...
         dec_out, dec_attn = self.model.decoder(
```
Comment (francoishernandez) [inline eole/predict/inference.py:681]: Why aren't we calling `_clear_cache` here (unlike in generator.py)?

File: eole/predict/translator.py [hunk 6]
```diff
@@ -141,9 +140,12 @@ def _run_encoder(self, batch):
         src = batch["src"]
         src_len = batch["srclen"]
         batch_size = len(batch["srclen"])
-        mask = sequence_mask(src_len)
         emb = self.model.src_emb(src)
-        enc_out, enc_final_hs = self.model.encoder(emb, mask)
+        pad_mask = src.eq(self._src_pad_idx).unsqueeze(1)  # [B, 1, T_src]
+        position_embeddings = self.model.rope.update(src.size(1), step=0)
```
Comment (francoishernandez) [inline eole/predict/translator.py:145]: Same question as inference/generator

File: eole/tests/test_model_lm/config.json [hunk 7]
```diff
@@ -10,7 +10,7 @@
       "layers": 2,
       "heads": 2
     },
-    "architecture": "transformer_lm",
+    "architecture": "custom",
```
Comment (francoishernandez) [inline eole/tests/test_model_lm/config.json:13]: Why is this needed?

File: eole/train_single.py [hunk 8]
```diff
@@ -188,6 +193,7 @@ def main(config, device_id):
     )
 
     if config.training.torch_compile:
+        torch._dynamo.config.cache_size_limit = 16
```
Comment (francoishernandez) [inline eole/train_single.py:196]: Should this be configurable?

**Review by francoishernandez — COMMENTED** [review]
Nice clean-up for the new year!
Quite a few comments not necessarily all relevant.

File: .github/workflows/push.yml [hunk 9]
```diff
@@ -146,15 +146,15 @@ jobs:
...
+    - name : Testing architecture transformer training w/ validation with dynamic scoring and rotary
       run: |
         python3 eole/bin/main.py train \
           -config eole/tests/data/data.yaml \
           -src_vocab /tmp/eole.vocab.src \
           -tgt_vocab /tmp/eole.vocab.tgt \
           -src_vocab_size 1000 \
           -tgt_vocab_size 1000 \
-          -model '{"architecture": "transformer", "layers": 4, "heads": 2, "hidden_size": 16, "transformer_ff": 64, "embeddings": {"word_vec_size": 16, "position_encoding_type": "Rotary"}}' \
+          -model '{"architecture": "transformer", "layers": 4, "heads": 2, "hidden_size": 16, "transformer_ff": 64, "rope_config": {}, "embeddings": {"word_vec_size": 16, "position_encoding_type": "Rotary"}}' \
```
[the remark this answers is not in the document]
Comment (vince62s) [reply .github/workflows/push.yml:157]: I can add the test in build_blocks() or maybe we can set the default value of rope_config to {} or there could be some pydantic logic but for this one not sure how/where to put it because the model is not built yet.

File: eole/decoders/ensemble.py [hunk 10]
```diff
@@ -164,9 +166,9 @@ class EnsembleModel(EncoderDecoderModel):
     """Dummy EncoderDecoderModel wrapping individual real EncoderDecoderModels."""
 
     def __init__(self, models, raw_probs=False):
-        src_emb = EnsembleSrcEmb(model.src_emb for model in models)
+        src_emb = EnsembleSrcEmb([model.src_emb for model in models])
```
[the remark this answers is not in the document]
Comment (vince62s) [reply eole/decoders/ensemble.py:169]: because we need a list and not a generator to access the first element

File: eole/decoders/transformer_decoder.py [hunk 11]
```diff
@@ -112,7 +111,7 @@ def _forward(
                 enc_out,
                 enc_out,
                 norm_layer_in,
-                mask=src_pad_mask,
+                attn_mask=~src_pad_mask,
```
[the remark this answers is not in the document]
Comment (vince62s) [reply eole/decoders/transformer_decoder.py:113]: it is way more homogeneous now:
pad_mask = 0 we don't mask, 1 we do mask
attn_mask = 0 we don't attend, 1 we do attend
this has been also aligned to the pytorch taxonomy

File: eole/decoders/transformer_decoder.py [hunk 12]
```diff
@@ -169,6 +169,7 @@ def __init__(
         self.layer_norm = LayerNorm[model_config.layer_norm](
             model_config.hidden_size, eps=model_config.norm_eps
         )
+        self._clear_cache()
```
[the remark this answers is not in the document]
Comment (vince62s) [reply eole/decoders/transformer_decoder.py:171]: the starting point was to avoid a test (step==None) at each iteration which is useless.
But ok, disable_cache probably a better wording

File: eole/inputters/text_utils.py [hunk 13]
```diff
@@ -180,7 +183,14 @@ def tensorify(vocabs, minibatch, device, left_pad=False):
         ]
     padidx = vocabs["src"][vocabs["specials"].get("pad_token", DefaultTokens.PAD)]
     tbatchsrc = pad_sequence(tbatchsrc, batch_first=True, padding_value=padidx)
-
+    """
+    This removes some recompiles in torch.dynamo, but slows down and make inference tricky
```
[the remark this answers is not in the document]
Comment (vince62s) [reply eole/inputters/text_utils.py:187]: It's just a note for myself. when using torch.compile() it spits out some recompiles in logs, reason being that the seq len is not a multiple of 8. BUT the strange thing is that if we use a context manager just before the scaled_dot_product_attention in MHA then the recompiles disappear. I need to investigate with Pytorch.

File: eole/models/model.py [hunk 14]
```diff
@@ -842,33 +863,39 @@ def __init__(self, **kwargs):
     def build_blocks(cls, model_config, vocabs, running_config=None):
         tgt_emb = build_tgt_emb(model_config, vocabs, running_config=running_config)
         decoder = build_decoder(model_config, running_config=running_config)
+        if (
+            model_config.embeddings.position_encoding_type
+            == PositionEncodingType.Rotary
+        ):
+            rope = RotaryPosition(model_config)
+        else:
+            rope = NoOpPosition()
```
[the remark this answers is not in the document]
Comment (vince62s) [reply eole/models/model.py:872]: by calling a super.build_blocks() ?

File: eole/predict/generator.py (hunk 1 above)
Comment (vince62s) [reply eole/predict/generator.py:181]: because when standard step decoding you need the cache

File: eole/predict/generator.py (hunk 2 above)
Comment (vince62s) [reply eole/predict/generator.py:182]: no, it happens after but would may need a super to something that has a different name:
def _predict_batch_with_strategy(self, batch, decode_strategy, left_pad=True):
def _translate_batch_with_strategy(self, batch, decode_strategy):
so for another PR

File: eole/predict/inference.py (hunk 3 above)
Comment (vince62s) [reply eole/predict/inference.py:665]: not anymore, it's just that as an intermediary change I tried to get rid of sequence_mask() but the tiling thing prevented me to do so.

File: eole/predict/inference.py (hunk 4 above)
Comment (vince62s) [reply eole/predict/inference.py:680]: making things 100% callable is way better for torch.compile (vs "if" everywhere, same logic as removing all of these in transformer_decoder / transformer_lm_decoder)

File: eole/predict/inference.py (hunk 5 above)
Comment (vince62s) [reply eole/predict/inference.py:681]: you're mistaken. we don't disable the cache when step decoding. We disable it when using the one-pass for estimation

File: eole/tests/test_model_lm/config.json (hunk 7 above)
Comment (vince62s) [reply eole/tests/test_model_lm/config.json:13]: it is not "NEEDED" but it's clearer since all settings are set in decoder and not at model level.
This appeared when I tried a further refactoring not duplicating all settings.

File: eole/train_single.py (hunk 8 above)
Comment (vince62s) [reply eole/train_single.py:196]: I would leave it this way for now but maybe in the future.
